Makefile

```diff
@@ -0,0 +1,194 @@
+REPO := logdna-agent-v2
+
+# The image repo and tag can be modified e.g.
+# `make build RUST_IMAGE=docker.io/rust:1.42.0`
+RUST_IMAGE_REPO ?= docker.io/rust
+RUST_IMAGE_TAG ?= 1.42
+RUST_IMAGE ?= $(RUST_IMAGE_REPO):$(RUST_IMAGE_TAG)
+RUST_IMAGE := $(RUST_IMAGE)
+
+HADOLINT_IMAGE_REPO ?= hadolint/hadolint
+HADOLINT_IMAGE_TAG ?= v1.8.0
+HADOLINT_IMAGE ?= $(HADOLINT_IMAGE_REPO):$(HADOLINT_IMAGE_TAG)
+HADLINT_IMAGE := $(HADOLINT_IMAGE)
+
+DOCKER := docker
+DOCKER_RUN := docker run --rm -w /build -v $(shell pwd):/build:Z
+DOCKER_PRIVATE_IMAGE := us.gcr.io/logdna-k8s/logdna-agent-v2
+DOCKER_PUBLIC_IMAGE := docker.io/logdna/logdna-agent
+DOCKER_IBM_IMAGE := icr.io/ext/logdna-agent
+
+VCS_REF := $(shell git rev-parse --short HEAD)
+VCS_URL := https://github.com/logdna/$(REPO)
+BUILD_TIMESTAMP := $(shell date -u +'%Y-%m-%dT%H:%M:%SZ')
+BUILD_VERSION := $(shell sed -nE "s/^version = \"(.+)\"\$$/\1/p" bin/Cargo.toml)
+BUILD_TAG := $(VCS_REF)
+
+MAJOR_VERSION := $(shell echo $(BUILD_VERSION) | cut -s -d. -f1)
+MINOR_VERSION := $(shell echo $(BUILD_VERSION) | cut -s -d. -f2)
+PATCH_VERSION := $(shell echo $(BUILD_VERSION) | cut -s -d. -f3 | cut -d- -f1)
+BETA_VERSION := $(shell echo $(BUILD_VERSION) | cut -s -d- -f2 | cut -s -d. -f2)
+ifeq ($(BETA_VERSION),)
+	BETA_VERSION := 0
+endif
+
+ENTRYPOINT := ./docker/entrypoint.sh $(shell id -u) $(shell id -g) "$(shell uname)" "/build"
+
+ifeq ($(ALL), 1)
+	CLEAN_TAG := *
+else
+	CLEAN_TAG := $(BUILD_TAG)
+endif
+
+CLEAN_DOCKER_IMAGES = if [[ ! -z "$(shell docker images -q $(1))" ]]; then docker images -q $(1) | xargs docker rmi -f; fi
+
+PULL ?= 1
+ifeq ($(PULL), 1)
+	PULL_OPTS := --pull
+else
+	PULL_OPTS :=
+endif
+
+CHANGE_VERSION = awk '{sub(/^version = ".+"$$/, "version = \"$(1)\"")}1' bin/Cargo.toml >> bin/Cargo.toml.tmp && mv bin/Cargo.toml.tmp bin/Cargo.toml
+
+REMOTE_BRANCH := $(shell git branch -vv | awk '/^\*/{split(substr($$4, 2, length($$4)-2), arr, "/"); print arr[2]}')
+
+.PHONY:build
+build: ## Build the agent
+	$(DOCKER_RUN) $(RUST_IMAGE) $(ENTRYPOINT) "cargo build"
+
+.PHONY:build-release
+build-release: ## Build a release version of the agent
+	$(DOCKER_RUN) $(RUST_IMAGE) $(ENTRYPOINT) "cargo build --release && strip ./target/release/logdna-agent"
+
+.PHONY:test
+test: ## Run unit tests
+	$(DOCKER_RUN) $(RUST_IMAGE) $(ENTRYPOINT) "cargo test"
+
+.PHONY:clean
+clean: ## Clean all artifacts from the build process
+	$(DOCKER_RUN) $(RUST_IMAGE) $(ENTRYPOINT) "cargo clean"
+
+.PHONY:clean-docker
+clean-docker: ## Cleans the intermediate and final agent images left over from the build-image target
+	@# Clean any agent images, left over from the multi-stage build
+	$(call CLEAN_DOCKER_IMAGES,$(REPO):$(CLEAN_TAG))
```

Comment:
this really could just be

```
	-docker images -q $(1) | xargs docker rmi -f
```

Reply:
I can move the `CLEAN_DOCKER_IMAGES` statement down to this target and drop the `$(call ...)` syntax, but I think I still need to keep the
```bash
if [[ ! -z "$(shell docker images -q $(1))" ]]; then
```

because on my Linux box it complains when the image is not there. The behavior is different on Mac (seems like the pipe silently exits on no data?) . I suppose it's a matter of taste, but I wouldn't want the clean command to fail because there's nothing to clean.

Reply:
The starting `-` prevents it from failing